src/lumps/textureLibrary.ts

```diff
@@ -0,0 +1,135 @@
+import {WADFlat} from "@src/lumps/doom/flat";
+import {WADTexture, WADTextures} from "@src/lumps/doom/textures";
+import {WADFileList} from "@src/wad/fileList";
+import {WADLump} from "@src/wad/lump";
+
+// Flats and walls use different texture sets.
+export enum TextureSet {
+    Walls,
+    Flats,
+}
+
+export class TextureLibrary {
+    // The textures themselves
+    protected textures: {
+        [TextureSet.Walls]: {[name: string]: WADTexture | null};
+        [TextureSet.Flats]: {[name: string]: WADFlat | null};
+    };
+    // And whether or not each one is transparent
+    protected transparent: {
+        [TextureSet.Walls]: {[name: string]: boolean | null};
+        [TextureSet.Flats]: {[name: string]: boolean | null};
+    };
+    // RGBA pixel buffers created from each texture
+    protected rgba: {
+        [TextureSet.Walls]: {[name: string]: Buffer | null};
+        [TextureSet.Flats]: {[name: string]: Buffer | null};
+    };
+    public fileList: WADFileList;
+    constructor(fileList: WADFileList){
+        this.fileList = fileList;
+        this.textures = {
+            [TextureSet.Walls]: {},
+            [TextureSet.Flats]: {},
+        };
+        this.transparent = {
+            [TextureSet.Walls]: {},
+            [TextureSet.Flats]: {},
+        };
+        this.rgba = {
+            [TextureSet.Walls]: {},
+            [TextureSet.Flats]: {},
+        };
+    }
+    protected isFlatMarker(lump: WADLump): boolean {
+        const name = lump.name;
+        return (
+            name === "F1_START" || name === "F2_START" || name === "F3_START" ||
+            name === "F1_END" || name === "F2_END" || name === "F3_END" ||
+            lump.dataLength === 0
+        );
+    }
+    get(name: string, set: TextureSet): WADTexture | WADFlat | null {
+        // Get a texture, or add it to the library if it has not already been added
+        // Already indexed, so return it immediately
+        if(this.textures[set][name] !== undefined){
+            // A texture could missing from the texture lists and flat collections.
+            console.log(`Using cached ${TextureSet[set]}[${name}]`);
+            return this.textures[set][name];
+        }
+        console.log(`${TextureSet[set]}[${name}] is not in the library.`);
+        // Wall textures are defined in TEXTUREx list entries
+        if(set === TextureSet.Walls){
+            for(let textureListIndex = 1; textureListIndex <= 3; textureListIndex++){
+                const listName = `TEXTURE${textureListIndex}`;
+                const listLump = this.fileList.map.get(listName);
+                if(listLump){
+                    const texList = WADTextures.from(listLump);
+                    const texture = texList.getTextureByName(name);
+                    if(texture){
+                        this.textures[set][name] = texture;
+                        this.transparent[set][name] = texture.isTransparent(this.fileList);
```

Comment:
I am open to discussion on this point, but:

Instead of having an `isTransparent` method that in most cases is going to be already doing all the work of `getPixelDataRGBA` just to get the answer to the question, I would place this line after L72 (the very next line where `this.rgba[set][name]` is assigned) and change it to something like `this.transparent[set][name] = this.hasTransparency(this.rgba[set][name])`.

The hypothetical `this.hasTransparency(pixelData: Buffer)` would enumerate every fourth byte (i.e. the alpha channel byte) and check if `pixelData.readUInt8(3 + pixelIndex) !== 0xFF` for `pixelIndex` from 0 until `pixelData.length / 4`.

Reply:
This change, if made, would include removing the `isTransparent` methods from other classes. I would be open to an optimization like `canHaveTransparency()` that is used to circumvent the `hasTransparency` check for graphics like flats, or BMP image data without an alpha channel, where enumerating all the pixels in a `getPixelDataRGBA` result buffer is pointless. Though I wouldn't expect this to be a bottleneck right now and so I wouldn't recommend adding such a method as part of this PR.